app/src/main/java/com/example/pps_tudai/mvp/presenter/StepsCounterPresenter.java

```diff
@@ -1,37 +1,64 @@
 package com.example.pps_tudai.mvp.presenter;
 
+import android.annotation.SuppressLint;
 import android.hardware.Sensor;
 import android.hardware.SensorEvent;
 import android.hardware.SensorEventListener;
 import android.hardware.SensorManager;
-
+import android.location.Location;
 import com.example.pps_tudai.data.entities.entity.User;
 import com.example.pps_tudai.mvp.model.StepsCounterModel;
 import com.example.pps_tudai.mvp.view.StepsCounterView;
-
+import com.example.pps_tudai.utils.Tools;
+import com.google.android.gms.location.FusedLocationProviderClient;
+import com.google.android.gms.location.LocationCallback;
+import com.google.android.gms.location.LocationRequest;
+import com.google.android.gms.location.LocationResult;
+import com.google.android.gms.tasks.OnSuccessListener;
+import com.google.android.gms.tasks.Task;
 import static com.example.pps_tudai.utils.IntUtils.ZERO;
 import static com.example.pps_tudai.utils.StringUtils.FALSE;
 import static com.example.pps_tudai.utils.StringUtils.TRUE;
 
 public class StepsCounterPresenter implements SensorEventListener {
 
     private static final int ONE = 1;
+    private final static int INTERVAL = 10000;
+    private final static int FAST_INTERVAL = 5000;
+    private final static float MARGIN_DISTANCE = 3;
+
     private final StepsCounterView counterView;
     private final StepsCounterModel counterModel;
     private User user;
     private SensorManager sensorManager;
     private Sensor sensor;
 
-    public StepsCounterPresenter(StepsCounterView counterView, StepsCounterModel counterModel, int id, SensorManager sensorManager) {
+    private FusedLocationProviderClient fusedLocationClient;
+    private LocationRequest locationRequest;
+    private LocationCallback locationCallback;
+
+    public StepsCounterPresenter(StepsCounterView counterView, StepsCounterModel counterModel, int id,
+                                 SensorManager sensorManager, FusedLocationProviderClient fusedLocationClient) {
         this.counterView = counterView;
         this.counterModel = counterModel;
-        this.init(id);
         this.sensorManager = sensorManager;
+        this.fusedLocationClient = fusedLocationClient;
+        this.init(id);
     }
 
+    @SuppressLint("MissingPermission")
```

Comment:
This is not a good way to check the location permission, please add the current condition.